### PR TITLE
some fixes related to SparseLinearComponent

### PR DESCRIPTION
**Pull request by hhadian**
Add Transpose() for SparseMatrix
More efficient CopyFromMat for SparseMatrix
Fix kUpdatableComponent and kBackpropNeedsInput bug
Set kPropagateAdds flag and modify Propagate(): out->AddMatSmat(1.0, in, kNoTrans, linear_params_, kTrans, 1.0); 

I'll prepare the rest of the fixes in the next commit. Also I'll implement transpose for GPU ASAP.

### CODE REVIEW

File: src/matrix/sparse-matrix.cc [hunk 1]
```diff
@@ -336,6 +336,22 @@ Real SparseMatrix<Real>::FrobeniusNorm() const {
   return std::sqrt(squared_sum);
 }
 
+template<typename Real>
+void SparseMatrix<Real>::Transpose() {
+  if(rows_.empty())
+    return;
+  std::vector<SparseVector<Real> > trows(NumCols(), SparseVector<Real>(NumRows()));
```
Comment (danpovey) [inline src/matrix/sparse-matrix.cc]: be aware of 80-char line limit; and use better name e.g. trans_rows.

**Comment by danpovey** [comment]
merging now- any fixes can  be merged later.

File: src/matrix/sparse-matrix.cc (hunk 1 above)
Comment (hhadian) [inline src/matrix/sparse-matrix.cc]: Ok. Certainly. I wanted to prepare it quickly; It will never happen again.

**Comment by danpovey** [comment]
oh- just noticed I didn't merge this already.  will do it now.

**Pull request merged**